### PR TITLE
Feat/waiver request

### PR DESCRIPTION
**Pull request by acpcreation**
More in depth testing and feature expansion for full coverage for supported platforms.

### CODE REVIEW

File: images/Sonatype-platform-icon.png
Comment (madpah) [inline images/Sonatype-platform-icon.png:1]: Why has this file been added?

File: README.md [hunk 1]
```diff
@@ -79,6 +103,7 @@ The URL to add is `http://localhost:3000/webhook`. Don't forget to substitue the
 
 Supported WebHook Events currently are:
 - Application Evaluation
+- Violation Alert
```
Comment (madpah) [inline README.md]: This is a new Webhook, but no additional TYPE has been added?

File: src/handlers/jira.ts [hunk 2]
```diff
@@ -108,9 +106,88 @@ export class JiraHandler extends BaseHandler {
...
+                                },
+                                {
+                                    "type": "text",
+                                    "text": "\nView Evaluation in Sonatype IQ",
+                                    "marks": [
+                                        {
+                                            "type": "link",
+                                            "attrs": {
+                                                "href": getIqUrlForApplicationEvaluation(payload),
+                                                "title": "IQ URL"
```
Comment (madpah) [inline src/handlers/jira.ts]: Should this not be more on brand? Sonatype Lifecycle?

File: src/handlers/jira.ts [hunk 3]
```diff
@@ -108,9 +106,88 @@ export class JiraHandler extends BaseHandler {
...
+                                        }
+                                    ]
+                                },
+                                {
+                                    "type": "text",
+                                    "text": `\n\t- Affected Components:\t ${payload.applicationEvaluation.affectedComponentCount} \n\t- Critical Issues:\t ${payload.applicationEvaluation.criticalComponentCount} \n\t- Severe Issues:\t ${payload.applicationEvaluation.severeComponentCount} \n\t- Moderate Issues:\t ${payload.applicationEvaluation.moderateComponentCount} \n Evaluation Date: ${payload.applicationEvaluation.evaluationDate} \nStage: ${payload.applicationEvaluation.stage} \nOutcome: ${payload.applicationEvaluation.outcome}`,
+                                },
+                                {
+                                    "type": "text",
+                                    "text": "\nView Evaluation in Sonatype IQ",
```
Comment (madpah) [inline src/handlers/jira.ts]: Branding

File: src/handlers/jira.ts [hunk 4]
```diff
@@ -108,9 +106,88 @@ export class JiraHandler extends BaseHandler {
...
+                "description": {
+                    "type": "doc",
+                    "version": 1,
+                    "content": [
+                        {
+                            "type": "paragraph",
+                            "content": [
+                                {
+                                    "type": "text",
+                                    "text": "Sonatype Scan Result for "
```
Comment (madpah) [inline src/handlers/jira.ts]: Sonatype Platform?

File: src/handlers/slack.ts [hunk 5]
```diff
@@ -23,12 +23,48 @@ import { BaseHandler } from "./base";
...
+                        },
+                    ]
+                }, {
+                    "type": "actions",
+                    "elements": [
+                        {
+                            "type": "button",
+                            "text": {
+                                "type": "plain_text",
+                                "text": "Create Waiver in IQ"
```
Comment (madpah) [inline src/handlers/slack.ts]: Branding "IQ"

File: src/handlers/slack.ts [hunk 6]
```diff
@@ -23,12 +23,48 @@ import { BaseHandler } from "./base";
...
+        }
+
+        target.sendMessage(message).catch(err => console.error(`Slack Error ${err}: ${err.response.data}`))
     }
     
+    
     public handleApplicationEvaluation(payload: IqWebhookPayloadApplicationEvaluation, target: WebhookTarget): void {
         const message = {
-            "text": "TESTING",
+            "text": "New Sonatype IQ Evaluation Results",
```
Comment (madpah) [inline src/handlers/slack.ts]: IQ - branding

File: src/handlers/teams.ts [hunk 7]
```diff
@@ -14,19 +14,50 @@
...
-
 import template from "../templates/adaptive-card-default.json";
 import { getIqUrlForApplicationEvaluation } from "../helpers/iq";
 import { HandlerNotImplementedError } from "../error"
 
 export class TeamsHandler extends BaseHandler {
     
     public handleWaiverRequest(payload: IqWebhookPayloadWaiverRequest, target: WebhookTarget): void {
-        throw new HandlerNotImplementedError("Method not implemented.")
+        const color = "ff8400" // Orange
```
Comment (madpah) [inline src/handlers/teams.ts]: constants that may be reused should be declared in a more shared place - e.g.  `constants.ts` ?

File: src/handlers/teams.ts [hunk 8]
```diff
@@ -14,19 +14,50 @@
...
+                    "value": `[Violation Details](${payload.policyViolationLink})`
+                },{
+                    "name": "Request Note: ",
+                    "value": `\"${payload.comment}\"`
+                }],
+                "markdown": true
+            }],
+            "potentialAction": [ {
+                "@type": "OpenUri",
+                "name": "Create Waiver in IQ",
```
Comment (madpah) [inline src/handlers/teams.ts]: IQ - branding

File: src/helpers/iq.ts [hunk 9]
```diff
@@ -18,5 +18,12 @@ import { IqWebhookPayloadApplicationEvaluation } from "../types";
 import { IQ_SERVER_URL } from '..';
 
 export function getIqUrlForApplicationEvaluation(payload: IqWebhookPayloadApplicationEvaluation): string {
```
Comment (madpah) [inline src/helpers/iq.ts:20]: I would normalize the IQ SERVER URL in `index.ts` rather than in here (revert this change).

Comment (madpah) [reply src/helpers/iq.ts:20]: Could be written on one line:
`const IQ_SERVER_URL = (IQ_SERVER_URL.slice(-1) == "/" ? IQ_SERVER_URL : `${IQ_SERVER_URL}/`)`

File: src/temp/teams.js
Comment (madpah) [inline src/temp/teams.js:1]: DELETE THIS TEST FILE

**Review by madpah — CHANGES_REQUESTED** [review]
Hey @acpcreation - got some revisions dude

File: images/Sonatype-platform-icon.png
Comment (acpcreation) [reply images/Sonatype-platform-icon.png:1]: This was added to upload as the image for the Slack and Teams webhooks. I wanted to link to our branding library instead but it is for internal use only.  :\

Sounds like a good usecase to keep the image!

File: src/temp/teams.js
Comment (acpcreation) [reply src/temp/teams.js:1]: Done!

File: src/handlers/jira.ts (hunk 2 above)
Comment (acpcreation) [reply src/handlers/jira.ts]: Done!

File: src/handlers/jira.ts (hunk 3 above)
Comment (acpcreation) [reply src/handlers/jira.ts]: Done!

File: src/handlers/jira.ts (hunk 4 above)
Comment (acpcreation) [reply src/handlers/jira.ts]: Done!

File: src/handlers/slack.ts (hunk 6 above)
Comment (acpcreation) [reply src/handlers/slack.ts]: Done!

File: src/handlers/teams.ts (hunk 8 above)
Comment (acpcreation) [reply src/handlers/teams.ts]: done!

File: src/handlers/slack.ts (hunk 5 above)
Comment (acpcreation) [reply src/handlers/slack.ts]: done!

File: src/helpers/iq.ts (hunk 9 above)
Comment (acpcreation) [reply src/helpers/iq.ts:20]: ...you did this to me...

`export const IQ_SERVER_URL: string = (process.env.IQ_SERVER_URL.slice(-1) != "/") ? process.env.IQ_SERVER_URL+"/" : process.env.IQ_SERVER_URL`

File: src/handlers/teams.ts (hunk 7 above)
Comment (acpcreation) [reply src/handlers/teams.ts]: Done- and then some! `constants.ts line 27`

File: README.md (hunk 1 above)
Comment (acpcreation) [reply README.md]: You are right, violation alert actually has a different header. Will remove this for now. Should I update the screenshot as well?

File: README.md [hunk 10]
```diff
@@ -29,11 +29,35 @@ This project contains an example (working) Web Hook handler for [Sonatype Lifecy
 
 #### Create a local config.json
 
-Create a `config.json` file formatted like the provided `example.config.json`.
+Create a `config.json` file formatted like the provided `example.config.json` in the `examples/` directory.
 
 The `config.json` will allow you to configure multiple endpoints for a single message from Sonatype IQ. You can configure 1 or several message types.
```
Comment (madpah) [inline README.md]: Branding

File: README.md [hunk 11]
```diff
@@ -67,7 +91,7 @@ On Slack we need to create an app to listen for our Webhooks from Sonatype Lifec
 4. Toggle "On" the Activate Incoming Webhooks then click "Add New Webhook to Workspace"
 5. Select the channel or contact we want to forward the Webhook messages to
 6. Copy that Webhook URL - you'll need it later!
-7. You can also update the display information at the bottom of the *Basic Information* page with the Sonatype logo (The icon is in attached in the "images" directory)
+7. You can also update the display information at the bottom of the *Basic Information* page with the Sonatype logo (The icon is in attached in the `images/` directory)
```
Comment (madpah) [inline README.md]: You were going to remove this from SCM and refer to the branding site?

File: src/constants.ts [hunk 12]
```diff
@@ -16,10 +16,37 @@
 
 export enum IqWebhookEvent {
     APPLICATION_EVALUATION = 'iq:applicationEvaluation',
-    WAIVER_REQUEST = 'iq:waiverRequest'
+    WAIVER_REQUEST = 'iq:waiverRequest',
+    // POLICY_ALERT = 'iq:policyAlert' // Violation Alert
```
Comment (madpah) [inline src/constants.ts]: Don't commit commented out code.

File: src/constants.ts [hunk 13]
```diff
@@ -16,10 +16,37 @@
...
 }
 
 export function getEnumKeyByEnumValue(myEnum, enumValue) {
     let keys = Object.keys(myEnum).filter(x => myEnum[x] == enumValue);
     return keys.length > 0 ? keys[0] : null;
+}
+
+export function getIQColor(color){
```
Comment (madpah) [inline src/constants.ts]: No need for this to be a function - it should just be a bunch of `const` exports - e.g.:

`export const COLOR_YELLOW = '#ffc233'`

File: src/handlers/teams.ts [hunk 14]
```diff
@@ -14,32 +14,64 @@
...
+
+
 import { AdaptiveCards } from "@microsoft/adaptivecards-tools";
 import { WebhookTarget } from "../WebHookTarget";
 import { IqWebhookPayloadApplicationEvaluation, IqWebhookPayloadWaiverRequest } from "../types";
 import { BaseHandler } from "./base";
-
 import template from "../templates/adaptive-card-default.json";
 import { getIqUrlForApplicationEvaluation } from "../helpers/iq";
+import { getIQColor } from "../constants";
```
Comment (madpah) [inline src/handlers/teams.ts]: As above - use constants.

File: src/helpers/iq.ts (hunk 9 above)
Comment (madpah) [reply src/helpers/iq.ts:20]: ?

File: src/index.ts [hunk 15]
```diff
@@ -29,10 +29,7 @@ require('dotenv').config()
 const app: Express = express()
 app.use(express.json());
 
-/**
- * Fill in your incoming webhook url.
- */
-export const IQ_SERVER_URL: string = process.env.IQ_SERVER_URL
+export const IQ_SERVER_URL: string = (process.env.IQ_SERVER_URL.slice(-1) != "/") ? process.env.IQ_SERVER_URL+"/" : process.env.IQ_SERVER_URL
```
Comment (madpah) [inline src/index.ts]: Probably shouldn't be exported here - it should be passed into methods requiring it to ensure they remain isolated and therefore testable.

**Review by madpah — CHANGES_REQUESTED** [review]
Bunch of changes advised and requested please @acpcreation

File: src/constants.ts (hunk 12 above)
Comment (acpcreation) [reply src/constants.ts]: Done!

File: images/Sonatype-platform-icon.png
Comment (madpah) [inline images/Sonatype-platform-icon.png:1]: This is still committed!

File: src/constants.ts [hunk 16]
```diff
@@ -16,10 +16,17 @@
 
 export enum IqWebhookEvent {
     APPLICATION_EVALUATION = 'iq:applicationEvaluation',
-    WAIVER_REQUEST = 'iq:waiverRequest'
+    WAIVER_REQUEST = 'iq:waiverRequest',
```
Comment (madpah) [inline src/constants.ts]: Why add a trailing comma?

File: src/handlers/teams.ts [hunk 17]
```diff
@@ -14,32 +14,63 @@
...
-        throw new HandlerNotImplementedError("Method not implemented.")
+        const message = {
+            "@type": "MessageCard",
+            "@context": "http://schema.org/extensions",
+            "themeColor": COLOR_ORANGE,
+            "summary": "New Sonatype Platform Policy Waiver Request",
+            "sections": [{
+                "activityTitle": "New Sonatype Platform Policy Waiver Request",
+                "activitySubtitle": `Requested By: ${payload.initiator}`,
+                // "activityImage": "https://teamsnodesample.azurewebsites.net/static/img/image5.png",
```
Comment (madpah) [inline src/handlers/teams.ts]: Remove commented code

**Review by madpah — CHANGES_REQUESTED** [review]
3 things to fix before we can merge.

**Review by madpah — APPROVED** [review]
